gpt_json/truncation.py

```diff
@@ -0,0 +1,86 @@
+import random
+from typing import Callable
+
+import tiktoken
+
+from gpt_json.models import GPTModelVersion, VariableTruncationMode
+
+
+def tokenize(text: str, model: GPTModelVersion) -> list[int]:
+    enc = tiktoken.encoding_for_model(model)
+    return [tok for tok in enc.encode(text)]
+
+
+def decode(tokens: list[int], model: GPTModelVersion) -> str:
+    enc = tiktoken.encoding_for_model(model)
+    return enc.decode(tokens)
+
+
+def num_tokens_from_messages(
+    messages: list[dict[str, str]], model: GPTModelVersion
+) -> int:
+    """Returns the number of tokens used by a list of messages.
+    NOTE: in the future, there may be structural changes to how messages are converted into content
```

Comment:
Nice updated doc, makes more sense now. I think it's still worth a bit more explicitness about our hard-coded strategy here, since there's a lot of numerical arithmetic that is tied pretty explicitly to a given architecture / message passing strategy.

Perhaps you can refactor this function like this (pseudocode follows):

```python
MODEL_MESSAGE_MARKUP = {
  GPTModelVersion.GPT_4: gpt_markup_v1,
  GPTModelVersion.GPT_3_5: gpt_markup_v1,
}

def gpt_markup_v1(...):
    # existing logic

def num_tokens_from_messages(
    messages: list[dict[str, str]], model: GPTModelVersion
) -> int:
    if model not in MODEL_MESSAGE_MARKUP:
        raise NotImplementedError()
    return MODEL_MESSAGE_MARKUP[model](messages, model)
```

And then in test cases validate that we've been explicit about choosing a message markup strategy for every model we support:

```python
@pytest.mark.parameterize("model", [*GPTModelVersion.values()])
def test_num_tokens(model):
    num_tokens_from_messages(messages, model)
```